web/src/providers/Test/Test.provider.tsx

```diff
@@ -0,0 +1,125 @@
+import {noop} from 'lodash';
+import {createContext, useCallback, useContext, useMemo, useState} from 'react';
+import {useGetTestByIdQuery, useGetTestVersionByIdQuery} from 'redux/apis/TraceTest.api';
+import {TDraftTest, TTest} from 'types/Test.types';
+import VersionMismatchModal from 'components/VersionMismatchModal';
+import useTestCrud from './hooks/useTestCrud';
+import TestService from '../../services/Test.service';
+
+interface IContext {
+  onEdit(values: TDraftTest): void;
+  onRun(): void;
+  isLoading: boolean;
+  isError: boolean;
+  test: TTest;
+  latestTest: TTest;
+  isLatestVersion: boolean;
+  isEditLoading: boolean;
+}
+
+export const Context = createContext<IContext>({
+  onEdit: noop,
+  onRun: noop,
+  test: {} as TTest,
+  latestTest: {} as TTest,
+  isLoading: false,
+  isError: false,
+  isLatestVersion: true,
+  isEditLoading: false,
+});
+
+interface IProps {
+  testId: string;
+  version?: number;
+  children: React.ReactNode;
+}
+
+export const useTest = () => useContext(Context);
+
+const TestProvider = ({children, testId, version = 0}: IProps) => {
+  const [isVersionModalOpen, setIsVersionModalOpen] = useState(false);
+  const [draft, setDraft] = useState<TDraftTest>({});
+  const [state, setState] = useState<'edit' | 'run'>();
+  const {runTest, edit, isEditLoading} = useTestCrud();
+  const {
+    data: test,
+    isLoading: isCurrentLoading,
+    isError: isCurrentError,
+  } = useGetTestVersionByIdQuery({testId, version}, {skip: !version});
+  const {data: latestTest, isLoading: isLatestLoading, isError: isLatestError} = useGetTestByIdQuery({testId});
+
+  const isLatestVersion = useMemo(
+    () => Boolean(version) && version === latestTest?.version,
+    [latestTest?.version, version]
+  );
+  const isLoading = isLatestLoading || isCurrentLoading;
+  const isError = isLatestError || isCurrentError;
+  const currentTest = (version ? test : latestTest)!;
+
+  const onEdit = useCallback(
+    (values: TDraftTest) => {
+      if (isLatestVersion) edit(test!, values);
+      else {
+        setState('edit');
```

Comment:
It took me a o couple of seconds to understand what was `state` - `setState`. Super minor suggestion, but maybe status could be a better word

Reply:
I changed it to `action`